editor/components/colors/with-colors.js

```diff
@@ -14,6 +14,40 @@ import { withSelect } from '@wordpress/data';
  */
 import { getColorValue, getColorClass, setColorValue } from './utils';
 
+const memoizedGetColor = memoize(
+	( colors ) =>
+		memoize(
+			( colorName, customColorValue, colorContext ) => {
+				return {
+					name: colorName,
+					class: getColorClass( colorContext, colorName ),
+					value: getColorValue( colors, colorName, customColorValue ),
+				};
+			},
+			( colorName, customColorValue, colorContext ) =>
+				`${ colorName }-${ customColorValue }-${ colorContext }`
+		)
+);
+/**
+ * Even though, we don't expect setAttributes to change memoizing it is essential.
+ * If setAttributes is not memoized, each time memoizedSetColor is called:
+ * a new function reference is returned (even if setAttributes has not changed).
+ * This would make our memoized chain useless.
+ */
+const memoizedSetColor = memoize(
```

Comment:
3 nested `memoize` calls look suspicious in here. Maybe it's okey, just saying :)

Reply:
Also, why would we want to preserve cache references to old `setAttributes` ? Seems like a use-case for [`memoize-one`](https://www.npmjs.com/package/memoize-one) or a `WeakMap`-based cache like the `withWeakMapCache` proposed in #6395.

Reply:
Unfortunately, memoize-one would not fit our use case, as each block has its setAttributes function and we would need to cache it for this approach to work.

I had the idea in mind that when a block was deleted, we would stop referencing the memoized function which would make its cache clear. But unfortunately given that I made the memoize call a top-level function, that is not true. So when blocks are deleted, we are still catching its deleted setAttributes :( Using WeakMap caching it would make the logic correct without leaks I think.
But anyway I got mind bugged and this solution is complex and probably not something we should pursue.

At the time using memoize seemed the best possibility to solve the rerender problem being back-compatible, now it seems the best long-term solution is changing the API. I'm still a little bit curious about how the rerenders could be solved keeping the same API without a WeakMap (and the preservation of references in deleted blocks) it seems there is a simple way and I'm not seeing it but now it is more an academic curiosity and not something I would pursue.

Thank you for your reviews I will follow up with them to address this problem.